# Homework 4

## 1.  Are “deep” neural networks destined to be another bust just as perceptrons and expert systems were in the past, or are they really a breakthrough that will be used for years into the future.
In my opinion, deep neural networks are not destined to be a bust, although I do think they are currently overvalued. Machine learning and neural networks, from my perspective, have become almost marketing terms in today’s world. Many new businesses have sprung up saying they can fix any problem through the use of these deep neural network Ais, but a neural networks has limitations. The expectations of people of what computers can do are frequently too idealistic. Deep networks need good data first and foremost in order to work, and the quality (as well as variety) of that data decides how accurate a model is. Due to this, and the often ludicrous computational cost of deep learning, neural networks are not the solution for every problem. For instance, during the ongoing pandemic of today, I imagine many deep neural networks would have difficulty predicting many different things (for instance the economy direction) because this new world does not fit with any data the model has ever been trained on. Deep neural networks are biased against rare events that do not occur frequently in a dataset, even when the rare events that do not show up in the dataset are vital. This is ultimately one of the flaws with these networks in real world applications, it is impossible to provide a dataset that is perfect.

In the end, machine learning may well have a “dark age” in its future when people’s expectations meet reality. Over the history of AI, a continuous cycle is apparent where a technique gains popularity, cannot fully meet the expectations of the AI community, and is replaced by a new technique eventually. Both expert systems and perceptrons had their time in the sun, but eventually had their limits fully understood over time. This cycle will probably continue as people begin to learn the full limitations of deep neural networks and work is done to try to reach even further. In the future, I imagine that a new type of AI will come to fruition and become popular, perhaps even based on the ideas from deep neural networks. Deep neural networks will not become a bust per say (they have already shown they can do amazing things), but the limitations of deep neural networks will assuredly be tempered over time as more limitations are made well known. 

    


## 2. Hand-compute a single, complete back-propagation cycle. 
1. Fill in weights.

    $\begin{aligned}
    &\begin{bmatrix}
    w_{i_1,h_1} & w_{i_1,h_2} \\
    w_{i_2,h_1} & w_{i_2,h_2}
    \end{bmatrix}
    \leftarrow
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} \\
    &\begin{bmatrix}
    w_{h_1, o_1} \\ 
    w_{h_2, o_1} 
    \end{bmatrix}
    \leftarrow
    \begin{bmatrix}
    0.14 \\
    0.15
    \end{bmatrix}
    \end{aligned}$
    
2. Compute the output for one sample (XOR: `[1, 1]` &rarr; `0`).

    $\begin{aligned}
    o_j &= 
    \begin{bmatrix}
    1 & 1 \\ 
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
    0.14 \\
    0.15
    \end{bmatrix}
    \\ &=
    \begin{bmatrix}
    1 * 0.11 + 1 * 0.21 & 1 * 0.12 + 1 * 0.08
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
    0.14 \\ 
    0.15
    \end{bmatrix}
    \\ &=
    \begin{bmatrix}
    0.32 & 0.20
    \end{bmatrix}
    \cdot
    \begin{bmatrix}
    0.14 \\ 
    0.15 
    \end{bmatrix}
    \\ &=
    \begin{bmatrix}
    0.32 * 0.14 + 0.20 * 0.15
    \end{bmatrix}
    \\ &= 0.0748
    \end{aligned}
    \\
    $

3. Compute the error and the delta).

    $\begin{aligned}
    L_2Error &= (1 - 0.0748)^2 \\
    &= 0.005595 \\
    \Delta_{o_1} &= (0 - 0.0748) \\
    &= -0.0748 \\
    \end{aligned}$

4. Back-propagate updates back through the network, assuming: 
    $learning\_rate = 0.05$; 
    Identity activation functions for all nodes, meaning always 1 for a constant.
     
    $\begin{aligned}
    \begin{bmatrix}
    w_{h_1, o_1} \\ 
    w_{h_2, o_1}
    \end{bmatrix} &\leftarrow 
    \begin{bmatrix}
    0.14 \\ 
    0.15 
    \end{bmatrix} + 0.05 \cdot 
    \begin{bmatrix}
    0.21 \\ 
    0.08 
    \end{bmatrix} \cdot 1.0 \cdot -0.0748 \\\\
    &= 
    \begin{bmatrix}
    0.14 \\ 
    0.15 
    \end{bmatrix} + 
    \begin{bmatrix}
    0.05 * 0.21 * 1.0 * -0.0748 \\
    0.05 * 0.08 * 1.0 * -0.0748 
    \end{bmatrix} \\
    &= 
    \begin{bmatrix}
    0.14 \\ 
    0.15 
    \end{bmatrix} +
    \begin{bmatrix}
    -0.0007854 \\
    -0.0002992 
    \end{bmatrix} \\
    &=
    \begin{bmatrix}
    0.1392146 \\ 
    0.1497
    \end{bmatrix}
    \end{aligned}$

    $\begin{aligned}
    \begin{bmatrix}
    w_{i_1,h_1} & w_{i_1,h_2} \\ 
    w_{i_2,h_1} & w_{i_2,h_2}
    \end{bmatrix} &\leftarrow 
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} + 0.05 \cdot
    \begin{bmatrix}
    1 & 1 \\ 
    1 & 1
    \end{bmatrix} \cdot 1.0 \odot
    \begin{bmatrix}
    0.14 & 0.15 \\ 
    0.14 & 0.15
    \end{bmatrix} \cdot -0.0748 \\ &=
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} + 
    \begin{bmatrix}
    0.05 * 1 * 1.0 & 0.05 * 1 * 1.0 \\ 
    0.05 * 1 * 1.0 & 0.05 * 1 * 1.0 \\ 
    \end{bmatrix} \odot 
    \begin{bmatrix}
    0.14 * -0.0748 & 0.15 * -0.0748\\ 
    0.14 * -0.0748 & 0.15 * -0.0748 
    \end{bmatrix} \\ &=
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} + 
    \begin{bmatrix}
    0.05 & 0.05 \\ 
    0.05 & 0.05 
    \end{bmatrix} \odot 
    \begin{bmatrix}
    -0.010472 & -0.01122 \\
    -0.010472 & -0.01122
    \end{bmatrix} \\ &=
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} + 
    \begin{bmatrix}
    0.05 * -0.010472 & 0.05 * -0.01122 \\
    0.05 * -0.010472 & 0.05 * -0.01122
    \end{bmatrix} \\ &=
    \begin{bmatrix}
    0.11 & 0.12 \\
    0.21 & 0.08
    \end{bmatrix} +     
    \begin{bmatrix}
    -0.0005236 & -0.000561 \\
    -0.0005236 & -0.000561
    \end{bmatrix} \\ &= 
    \begin{bmatrix}
    0.1094764 & 0.119439 \\
    0.2094764 & 0.079439
    \end{bmatrix}  
    \end{aligned}$

## 3. fashion_mnist ConvNet

In [1]:
from keras import layers
from keras import models
from keras.datasets import fashion_mnist
from keras.utils import to_categorical, plot_model
import numpy as np
import matplotlib.pyplot as plt

model = models.Sequential()

# Configure a convnet with 1 layer of convolution, max pooling, and drop out.
model.add(layers.Conv2D(64, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))

# Add layers to flatten the 2D image and then do a 10-way classification.
model.add(layers.Flatten())
model.add(layers.Dense(128, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))
model.summary()

# Format image data
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)
train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255
test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

#train model
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
# (can add validation_split=.15 for validation set testig)
history = model.fit(train_images, train_labels, epochs=5, batch_size=128)

# check for overfitting using validation
# loss = history.history['loss']
# val_loss = history.history['val_loss']
# epochs = range(1, len(loss) + 1)
# plt.plot(epochs, loss, 'bo', label='Training loss')
# plt.plot(epochs, val_loss, 'b', label='Validation loss')
# plt.title('Training and validation loss')
# plt.xlabel('Epochs')
# plt.ylabel('Loss')
# plt.legend()
#plt.show()

#evaluate against the test data
model.evaluate(test_images, test_labels)

Using TensorFlow backend.


Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 26, 26, 64)        640       
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 13, 13, 64)        0         
_________________________________________________________________
flatten_1 (Flatten)          (None, 10816)             0         
_________________________________________________________________
dense_1 (Dense)              (None, 128)               1384576   
_________________________________________________________________
dense_2 (Dense)              (None, 10)                1290      
Total params: 1,386,506
Trainable params: 1,386,506
Non-trainable params: 0
_________________________________________________________________
Epoch 1/5
60000/60000 [==============================] - 39s 653us/step - loss: 0.4613 - accuracy: 0.8374

[0.2551331055045128, 0.9124000072479248]

While I tested using more conv layers as well as adding dropout, the best performance
I was able to get out of my network was using a convnet with 1 layer of convolution, max pooling, and then
using a dense layer of 128 nodes. This got my accuracy to 91.24%.